crates/runtime/src/libcalls.rs

```diff
@@ -137,3 +142,195 @@ pub unsafe extern "C" fn wasmtime_imported_memory32_size(
+    let instance = (&mut *vmctx).instance();
+    let dst_table = instance.get_defined_table(dst_table_index);
+    let src_table = instance.get_foreign_table(src_table_index);
+    if let Err(trap) = Table::copy(dst_table, src_table, dst, src, len, source_loc) {
+        raise_lib_trap(trap);
+    }
+}
+
+/// Implementation of `table.copy` when the destination table is imported
+/// and the source table is locally defined.
+#[no_mangle]
+pub unsafe extern "C" fn wasmtime_table_copy_imported_defined(
+    vmctx: *mut VMContext,
+    dst_table_index: u32,
+    src_table_index: u32,
+    dst: u32,
+    src: u32,
+    len: u32,
+    source_loc: u32,
+) {
+    let dst_table_index = TableIndex::from_u32(dst_table_index);
+    let src_table_index = DefinedTableIndex::from_u32(src_table_index);
+    let source_loc = ir::SourceLoc::new(source_loc);
+    let instance = (&mut *vmctx).instance();
+    let dst_table = instance.get_foreign_table(dst_table_index);
+    let src_table = instance.get_defined_table(src_table_index);
+    if let Err(trap) = Table::copy(dst_table, src_table, dst, src, len, source_loc) {
+        raise_lib_trap(trap);
+    }
+}
+
+/// Implementation of `table.copy` when both tables are imported.
+#[no_mangle]
+pub unsafe extern "C" fn wasmtime_table_copy_imported_imported(
+    vmctx: *mut VMContext,
+    dst_table_index: u32,
+    src_table_index: u32,
+    dst: u32,
+    src: u32,
+    len: u32,
+    source_loc: u32,
+) {
+    let dst_table_index = TableIndex::from_u32(dst_table_index);
+    let src_table_index = TableIndex::from_u32(src_table_index);
+    let source_loc = ir::SourceLoc::new(source_loc);
+    let instance = (&mut *vmctx).instance();
+    let dst_table = instance.get_foreign_table(dst_table_index);
+    let src_table = instance.get_foreign_table(src_table_index);
+    if let Err(trap) = Table::copy(dst_table, src_table, dst, src, len, source_loc) {
+        raise_lib_trap(trap);
+    }
+}
+
+/// Implementation of `table.init`.
+#[no_mangle]
+pub unsafe extern "C" fn wasmtime_table_init(
+    vmctx: *mut VMContext,
+    table_index: u32,
+    elem_index: u32,
+    dst: u32,
+    src: u32,
+    len: u32,
+    source_loc: u32,
+) {
+    let table_index = TableIndex::from_u32(table_index);
+    let source_loc = ir::SourceLoc::new(source_loc);
+    let elem_index = PassiveElemIndex::from_u32(elem_index);
+
+    let instance = (&mut *vmctx).instance();
+
+    if let Err(trap) = instance.table_init(table_index, elem_index, dst, src, len, source_loc) {
+        raise_lib_trap(trap);
+    }
+}
+
+/// Implementation of `elem.drop`.
+#[no_mangle]
+pub unsafe extern "C" fn wasmtime_elem_drop(vmctx: *mut VMContext, elem_index: u32) {
+    let elem_index = PassiveElemIndex::from_u32(elem_index);
+    let instance = (&mut *vmctx).instance();
+    instance.elem_drop(elem_index);
+}
+
+/// Implementation of `memory.copy` for locally defined memories.
+#[no_mangle]
+pub unsafe extern "C" fn wasmtime_memory_copy(
```

Comment:
Could this be renamed `wasmtime_defined_memory_copy`?

(or perhaps get merged with `wasmtime_imported_memory_copy`, I'm still not entirely sure why there's two separate functions)

Reply:
mostly because I was cribbing from how we do `memory.size` and `memory.grow`

I'll do the renaming in this PR and file an issue for merging imported/defined memory impls together. https://github.com/bytecodealliance/wasmtime/issues/985